src/jit/gentree.h

```diff
@@ -1909,15 +1916,55 @@ struct GenTree
     {
         return (gtFlags & GTF_REG_VAL) ? true : false;
     }
-    void SetInReg()
+    void SetInReg(bool value = true)
     {
-        gtFlags |= GTF_REG_VAL;
+        if (value == true)
+        {
+            gtFlags |= GTF_REG_VAL;
+        }
+        else
+        {
+            gtFlags &= ~GTF_REG_VAL;
+        }
     }
-
     regNumber GetReg() const
     {
         return InReg() ? gtRegNum : REG_NA;
     }
+
+#else // !LEGACY_BACKEND
+    // For the non-legacy backend, these are only used for dumping.
+    // The gtRegNum is only valid in LIR, but the dumping methods are not easily
+    // modified to check this.
+#ifdef DEBUG
+    bool InReg() const
+    {
+        return (GetRegTag() != GT_REGTAG_NONE) ? true : false;
+    }
+    regNumber GetReg() const
+    {
+        return (GetRegTag() != GT_REGTAG_NONE) ? gtRegNum : REG_NA;
+    }
+#endif
+
+    static bool IsContained(unsigned flags)
+    {
+        return ((flags & GTF_CONTAINED) != 0);
+    }
+
+    void SetAsContained()
+    {
+        gtFlags |= GTF_CONTAINED;
+    }
+
+    void ResetContained()
+    {
+        gtFlags &= ~GTF_CONTAINED;
+        gtLsraInfo.regOptional = false;
```

Comment:
So ... this is one step on the way. As I mentioned in https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/lsra-throughput.md#representing-containedness my original plan was to use a `GTF_TREE_ROOT` flag but then I realized that I should just make it an explicit `GTF_CONTAINED` flag. The next steps (in some order) are to:
* replace `gtLsraInfo.regOptional` with `REG_OPT` and `REG_OPT_USE` values for `gtRegNum`.
* move the setting of these into `Lowering`
* remove `gtLsraInfo` out of `GenTree`, since it can now be determined without any context from parent nodes